packages/aws-cdk/lib/cdk-toolkit.ts

```diff
@@ -266,8 +288,30 @@ export class CdkToolkit {
           });
         }
       }
+
       print('\n✨  Total time: %ss\n', formatTime(elapsedSynthTime + elapsedDeployTime));
-    }
+    };
+
+    const isStackUnblocked = (stack: cxapi.CloudFormationStackArtifact) =>
+      stack.dependencies
+        .map(({ id }) => id)
+        .filter((id) => !id.endsWith('.assets'))
+        .every((id) => !stacksAwaitingDeploy[id]);
+
+    const enqueueStackDeploys = async () => {
+      while (stacks[0] && isStackUnblocked(stacks[0])) {
```

Comment:
We're leaving some concurrency on the table here by only considering the first element from the list.

If we had dependencies:

```
C -> A
D -> B
```

And the list happened to toposort as `[A, B, C, D]`, then with this implementation even if `B` finishes we still can't start `D` because we are forced to wait for `C` to start first.

This can be changed to a `for` loop -- on every iteration, consider all stacks.

----

It should probably be a `forEach`, otherwise we might run afoul of the "[closing over a loop variable](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures#creating_closures_in_loops_a_common_mistake)" gotcha.

Reply:
Great catch! Updated to a `forEach` loop